[dotnet] [bidi] Modules as extensions

nvborisenko opened the pull request:
### **User description**
Contributes to https://github.com/SeleniumHQ/selenium/issues/15329

### 🔗 Related Issues
<!-- Example: Fixes #1234 or Closes #5678 -->
<!-- If the reason for this PR is not obvious, consider creating an issue for it first -->

### 💥 What does this PR do?
<!-- Describe what this change includes and how it works -->

### 🔧 Implementation Notes
<!--- Why did you implement it this way? -->
<!--- What alternatives to this approach did you consider? -->

### 💡 Additional Considerations
<!--- Are there any decisions that need to be made before accepting this PR? -->
<!--- Is there any follow-on work that needs to be done? (e.g., docs, tests, etc.) -->

### 🔄 Types of changes
<!-- ✂️ Please delete anything that doesn't apply -->
- Cleanup (formatting, renaming)
- Bug fix (backwards compatible)
- New feature (non-breaking change which adds functionality *and tests!*)
- Breaking change (fix or feature that would cause existing functionality to change)


___

### **PR Type**
Enhancement


___

### **Description**
- Refactor BiDi modules to use extension pattern

- Centralize JSON serialization configuration in BiDi class

- Replace individual module properties with generic AsModule method

- Update all module constructors to parameterless design


___

### Diagram Walkthrough


```mermaid
flowchart LR
  A["BiDi Class"] --> B["AsModule<T>() Method"]
  B --> C["ConcurrentDictionary<Type, Module>"]
  C --> D["Module.Create<T>()"]
  D --> E["Individual Modules"]
  A --> F["Centralized JsonSerializerOptions"]
  F --> G["BiDiJsonSerializerContext"]
  G --> E
```



<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>13 files</summary><table>
<tr>
  <td><strong>BiDi.cs</strong><dd><code>Centralize JSON context and implement module extensions</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-3f77d71f94e2a47ea90b74f6d37b5a5b94dc83b2f558b30ab070588b9028c2f6">+46/-129</a></td>

</tr>

<tr>
  <td><strong>Module.cs</strong><dd><code>Add factory method and JSON context properties</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-b45f63b7ea084aed81f19bba490ac80626f908cecea9448d35db3d4fbc663ddc">+22/-2</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>Broker.cs</strong><dd><code>Update command execution with JSON context parameter</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-5f7fc0e0b816bad29734139ccd27ebb0ff8809c8f49de7827eab7c6c7e8f9669">+22/-58</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>BrowserModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-8e4bcc00ef63cf737c6e26397ac40f1e18d960e4ff47589fee1bae1b9fe8cdc7">+10/-10</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>BrowsingContextModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-1f04192c50706a21f9dfdc1b79916855ccf2aba6c2c895811f158eeac915680c">+41/-41</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>SessionModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-31b1c3c5b298bcc68ac630c0af2fe9ecd548a737be8b5fa14574317f6eff3ff3">+6/-6</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>NetworkModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-ddd0c9f9565ffffc6a3ce489431b76319cc38c727006605f7dc6345b2c0f7c81">+25/-25</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>ScriptModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-da8cf630954e0732aa2b7f257aca8b49117eeceae343e373e2ba5914e639b6d2">+12/-12</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>LogModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-fdb6bcf97d10b7240dada7ced38218b43fcb0b8ba716b273dd9c60e17c286e3f">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>StorageModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-c2832a3991518e29f9a8f706cbe7e7c1d5ec4a007fbb57da280bb1be9fc7b2c2">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>InputModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-f98f384929541b30f9b87fb08f96ead9181d6b14a2353f5485ba985ed7cbd672">+4/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>EmulationModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-326495cc635fdb9686d337228f031a251b2aaed55960d7096b1345e0f1c267ff">+10/-11</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>WebExtensionModule.cs</strong><dd><code>Remove constructor parameter and use JsonContext</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/SeleniumHQ/selenium/pull/16392/files#diff-0eaa7ea3946fcd852a8db4e125d1bd98c674f8ce8489cc0279f452dca8f30a25">+3/-3</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

nvborisenko committed:
Dummy

nvborisenko commented:
@RenderMichael despite on these dummy changes, let's see how we can modify `JsonSerializerContext/Options` in runtime. I have no ideas, the best what I got: `System.InvalidOperationException : This JsonSerializerOptions instance is read-only or has already been used in serialization or deserialization.`

If you know please advise.

RenderMichael commented on dotnet/src/webdriver/BiDi/Module.cs:
```diff
@@ -21,7 +21,17 @@
-    protected Broker Broker { get; } = broker;
+    protected Broker Broker { get; private set; }
+
+    protected internal abstract void Initialize(Broker broker);
```
Do we want third-party code to be able to make their own modules? If so, we can make this `protected` instead of `protected internal` so others can override.

nvborisenko replied:
Yes, resolved.

nvborisenko commented:
I don't want to believe that injecting custom `TypeInfoResolver` is not possible.

RenderMichael commented on dotnet/src/webdriver/BiDi/Communication/Broker.cs:
```diff
@@ -205,21 +212,14 @@ private async Task<EmptyResult> ExecuteCommandCoreAsync<TCommand>(TCommand comma
 
-        var data = JsonSerializer.SerializeToUtf8Bytes(command, typeof(TCommand), _jsonSerializerContext);
-
+        var data = JsonSerializer.SerializeToUtf8Bytes(command, typeof(TCommand), _jsonSerializerOptions);
```
Now that we are passing in options instead of context, we can make this strongly typed

```suggestion
        var data = JsonSerializer.SerializeToUtf8Bytes(command, _jsonSerializerOptions);
```

RenderMichael commented on dotnet/src/webdriver/BiDi/Communication/Broker.cs:
```diff
@@ -108,7 +108,14 @@ internal Broker(BiDi bidi, Uri url)
 
-        _jsonSerializerContext = new BiDiJsonSerializerContext(jsonSerializerOptions);
+        // Add base BiDi generated context resolver; keep options mutable for module contexts
+        _jsonSerializerOptions.TypeInfoResolverChain.Add(BiDiJsonSerializerContext.Default);
```
We should factor out `BiDiJsonSerializerContext` to instead use the `Initialize` method instead

RenderMichael committed:
Merge branch 'trunk' into bidi-module-init

RenderMichael committed:
Refactor so each module controls its own JSON serialization

RenderMichael committed:
Remove commented code

nvborisenko commented:
Nice! Thank you.

Parameterless ctor for Module looks ugly. But we need something parameterless to be able to `new T()`. I think we can introduce `TModuleFactory`, and then `new TModuleFactory().Create(any args)`. It would be better, because we avoid `Initialize()` method at all, and move this logic into module's ctor.

RenderMichael commented:
If we targeted only the latest .NET versions, this would be a great place to have a static abstract member:

```csharp

interface ICreateModule<TModule> where TModule : ICreateModule<TModule>
{
    static abstract TModule Create(Broker broker);
}

public class MyModule : Module, ICreateModule<MyModule>
{
    private JsonSerializerContext _context;
    private MyModule(Broker broker, JsonSerializerOptions options) : base(broker)
    {
        _context = new MySerializerContext(broker.CreateOptions());
    }

    public static MyModule Create(Broker broker)
    {
        return new MyModule(broker);
    }
}
```

But alas, we cannot expose it the "best" way on modern .NET but a different way on older TFMs.

---



> I think we can introduce `TModuleFactory`, and then `new TModuleFactory().Create(any args)`. It would be better, because we avoid `Initialize()` method at all, and move this logic into module's ctor.

I don't see how that would work differently from the `Module.Create<TModule>(any args)` introduced here. Beyond the static abstract approach, or some reflection-based constructor duck-typing, I do not think there is another option.

RenderMichael committed:
Tighten flow for JSON configuration

nvborisenko commented:
I tried to implement different factory approaches, no luck (anyway consumer should provide factory func when calling, meaning this signature is exposed). Current approach looks better:

- Module parameterless `ctor`
- Module is enforced to provide some initialization (via `protected abstract`)

I will continue.

nvborisenko committed:
Hold json context per module

Forgot BrowserModule

Unnecessary Command/EmptyResult types in context?

Move BrowsingContext

Move Session

Move Storage

Move WebExtension

Move Script

Move Network

Move Log

Move Emulation

Move Input

Delete global json context

nvborisenko commented:
Now each module owns `JsonSerializerContext`, and I see 2 issues:

### **Performance degradation (memory)**

Before:
```
| Method                         | Mean     | Error   | StdDev  | Gen0     | Gen1     | Gen2     | Allocated |
|------------------------------- |---------:|--------:|--------:|---------:|---------:|---------:|----------:|
| SeleniumCaptureScreenshotAsync | 128.1 ms | 6.83 ms | 4.52 ms | 250.0000 | 250.0000 | 250.0000 |   1.65 MB |
```

After:
```
| Method                         | Mean     | Error   | StdDev  | Gen0     | Gen1     | Gen2     | Allocated |
|------------------------------- |---------:|--------:|--------:|---------:|---------:|---------:|----------:|
| SeleniumCaptureScreenshotAsync | 128.5 ms | 9.98 ms | 5.94 ms | 500.0000 | 500.0000 | 500.0000 |   2.31 MB |
```

### **Cross-module type referencing**

If we define json context per module, then we also have to define all types this module is referencing to. It is nightmare! We should track references somehow, and it is easy to make a mistake.

### How to move on?
I think we can use one big shared json context by all modules. External modules will define their own context.

nvborisenko committed:
Introduce AsModule method in BiDi

nvborisenko committed:
Revert to one big json context

Fix

nvborisenko committed:
Move ownership of json options to bidi

nvborisenko committed:
BiDi shared json context for built-in modules

nvborisenko committed:
Don't allow to modify json context from derived type

nvborisenko commented:
@RenderMichael I am confident with proposed changes, please review.

- `BiDi` is owner of `JsonOptions/Context`
- Built-in modules reuse `JsonContext`
- External module may create own `JsonContext` and propagate to commands

Edited: Module has access to `JsonContext`, meaning it can provide specific `JsonTypeInfo` for serialization process. It is beneficial, I already measured.

RenderMichael commented on dotnet/src/webdriver/BiDi/Module.cs:
```diff
@@ -18,10 +18,30 @@
+
+    protected virtual void Initialize(JsonSerializerOptions options) { }
+
+    internal static TModule Create<TModule>(BiDi bidi, Broker broker, JsonSerializerOptions jsonOptions, BiDiJsonSerializerContext context)
```
`bidi` is an unused parameter

nvborisenko replied:
True, just included it here to make it possible to create custom stateful json converters. Like we already do for `BrowsingContextConverter`.

RenderMichael commented on dotnet/src/webdriver/BiDi/BiDi.cs:
```diff
@@ -18,164 +18,81 @@
+
+    public Emulation.EmulationModule Emulation => AsModule<Emulation.EmulationModule>();
+
+    public TModule AsModule<TModule>() where TModule : Module, new()
```
The `AsModule<TModule>()` method seems like overkill. Is there any chance we could avoid lazy-loading the modules entirely?

nvborisenko replied:
I don't know. Let image we will have `bidi.AsPermissions()` extension method, then it internally calls `bidi.AsModule<PermisssionsModule>()`, which wants to cache.

Any particular concerns?

nvborisenko replied:
Mike, I have upcoming interesting PR: https://github.com/SeleniumHQ/selenium/pull/16397

Resolving this thread to move further. But in any case let's keep on track.

RenderMichael reviewed (approved):
Looks pretty good! A few comments, none of them blocking.
